verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,231 @@
+                        type |-> "CHMsgTry",
+                        stateProof |-> sProof,
+                        clientProof |-> cProof]
+       IN /\ outBuf' = Append(outBuf, replyMsg)
+          /\ store' = [store EXCEPT !.connection = newCon]
+
+
+(* Handles a "CHMsgTry" message.
+ *)
+HandleTryMsg(m) ==
+    (* The good-case path. *)
+    /\ \/ store.connection.state = "UNINIT"
+       \/ store.connection.state = "INIT" /\ CheckLocalParameters(m.parameters)
+    /\ Len(outBuf) < MaxBufLen
+    /\ LET newCon == [parameters |-> m.parameters, 
+                      state |-> "INIT"]
+           sProof == GetStateProof
+           cProof == GetClientProof
+           replyMsg == [parameters |-> FlipConnectionParameters(m.parameters),
+                        type |-> "CHMsgAck"]
+       IN /\ outBuf' = Append(outBuf, replyMsg)
+          /\ store' = [store EXCEPT !.connection = newCon]
+
+
+(* Handles a "CHMsgAck" message.
+ *)
+HandleAckMsg(m) ==
+    /\ \/ store.connection.state = "INIT"
+       \/ store.connection.state = "TRYOPEN"
+    /\ CheckLocalParameters(m.parameters)
+    /\ Len(outBuf) < MaxBufLen
+    /\ LET newCon == [parameters |-> m.parameters, 
+                      state |-> "INIT"]
+           sProof == GetStateProof
+           cProof == GetClientProof
+           replyMsg == [parameters |-> FlipConnectionParameters(m.parameters),
+                        type |-> "CHMsgAck"]
+       IN /\ outBuf' = Append(outBuf, replyMsg)
+          /\ store' = [store EXCEPT !.connection = newCon]
+
+
+(* Handles a "CHMsgConfirm" message.
+ *)
+HandleConfirmMsg(m) ==
+    UNCHANGED<<outBuf, store>>
+
+
+\* If MaxHeight is not yet reached, then advance the height of the chain. 
+AdvanceChainHeight ==
+    /\ store.height < MaxHeight
+    /\ store' = [store EXCEPT !.height = @ + 1]
+    /\ UNCHANGED <<outBuf, inBuf>>
+
+
+(* Generic action for handling any type of inbound message.
+    Expects as parameter a message.
+    Takes care of invoking priming the 'store' and any reply msg in 'outBuf'.
+ *)
+ProcessMsg(m) ==
+    /\ ValidConnectionParameters(m.parameters) = TRUE
+    /\ ValidMessageType(m.type) = TRUE
+    (* One of the following disjunctions will always enable, since
+        the message type is guaranteed to be valid in the THEN branch. *)
+    /\ \/ m.type = "CHMsgInit" /\ HandleInitMsg(m)
+       \/ m.type = "CHMsgTry" /\ HandleTryMsg(m)
+       \/ m.type = "CHMsgAck" /\ HandleAckMsg(m)
+       \/ m.type = "CHMsgConfirm" /\ HandleConfirmMsg(m)
+
+
+(***************************************************************************
+ Connection Handshake Module main spec.
+ ***************************************************************************)
+
+
+Init(chainID) ==
+    /\ store = [id |-> chainID,
+                height |-> 1,
+                connection |-> nullConnection,
+                client |-> nullClient]
+
+
+Next ==
+    LET m == Head(inBuf)
+    IN /\ inBuf /= <<>>        \* Enabled if we have an inbound msg.
```

Comment:
What is the reasoning behind modelling inBuf as a sequence?

Reply:
We wanted to permit multiple outstanding messages to each chain.

If we permit a single message (i.e., we model communication with a buffer/shared location), then we need to make sure that the environment (see `MaliciousNextEnv` action) never overwrites this location as long as the initialization message is there (see `InitEnv`), which is quite complicated. If the environment can overwrite the initialization message, then we get a deadlock, because the chains will never move out of the "UNINIT" state.